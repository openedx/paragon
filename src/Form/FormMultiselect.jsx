import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { ExpandMore, Close } from '../../icons';
import Button from '../Button/index';
import { Form } from '../index';
import Chip from '../Chip';

export const MULTISELECT_NO_OPTIONS_TEXT = 'No options';

const FormMultiselect = React.forwardRef(({
  floatingLabel,
  options,
  noOptionsText,
  selectedOptions,
  setSelectedOptions,
  errorText,
  disabled,
  variant,
}, ref) => {
  const defaultOptions = (options || []).filter(item => !selectedOptions.includes(item));
  const [isOpen, setIsOpen] = useState(false);
  const [isActiveInput, setIsActiveInput] = useState(false);
  const [optionItems, setOptionItems] = useState(defaultOptions);
  const [searchValue, setSearchValue] = useState('');
  const inputRef = useRef(null);
  const dropdownOptions = searchValue
    ? optionItems.filter(item => item.toLowerCase().includes(searchValue.toLowerCase()))
    : optionItems;

  const handleInputFocus = () => {
    setIsActiveInput(true);
    setIsOpen(true);
  };

  const handleSearch = (e) => {
    setSearchValue(e.target.value);
    setIsOpen(true);
  };

  const handleRemove = (option) => {
    setOptionItems([...optionItems, option]);
    setSelectedOptions(selectedOptions.filter(item => item !== option));
  };

  const handleReset = () => {
    setSelectedOptions([]);
    setOptionItems(options);
  };

  const handleSelect = (option) => {
    setOptionItems(optionItems.filter(item => item !== option));
    setSelectedOptions([...selectedOptions, option]);
    inputRef.current.focus();
    setSearchValue('');
  };

  return (
    <div className="pgn__form-multiselect" ref={ref}>
      <div
        className={classNames('pgn__form-multiselect-field', {
          error: !!errorText,
          disabled,
        })}
      >
        <div className="pgn__form-multiselect-field-wrapper">
          <div className={classNames('pgn__form-multiselect-field-label',
            isActiveInput || !!selectedOptions.length ? 'slide-up' : 'slide-down', {
              dark: variant === 'dark',
            })}
          >
            {floatingLabel}
          </div>
          {selectedOptions.map((item) => (
            <Chip
              iconAfter={Close}
              onClick={() => handleRemove(item)}
              className="pgn__form-multiselect-field-chip"
              key={item}
            >
              <span className="pgn__form-multiselect-field-chip-title">
                {item}
              </span>
            </Chip>
          ))}
          <Form.Control
            ref={inputRef}
            className="pgn__form-multiselect-search"
            type="text"
            value={searchValue}
            onChange={handleSearch}
            placeholder={(isActiveInput || selectedOptions.length) ? '' : floatingLabel}
            onFocus={handleInputFocus}
            onBlur={() => setIsActiveInput(false)}
          />
        </div>
        <div className="pgn__form-multiselect-button-group">
          {!!selectedOptions.length && (
          <Button
            className="pgn__form-multiselect-field-hide-btn"
            iconAfter={Close}
            onClick={handleReset}
          />
          )}
          <span className="pgn__form-multiselect-button-group-border" />
          <Button
            className="pgn__form-multiselect-field-show-btn"
            onClick={() => setIsOpen(!isOpen)}
            iconAfter={ExpandMore}
          />
        </div>
      </div>
      <span className={classNames('pgn__form-multiselect-field-error', {
        error: !!errorText,
      })}
      >
        {errorText}
      </span>
      <div className={classNames('pgn__form-multiselect-items', isOpen ? 'show' : 'none')}>
        {dropdownOptions.length ? dropdownOptions.map(item => (
          <button
            className={classNames('pgn__form-multiselect-item ')}
            key={`option-${item}`}
            type="button"
            onClick={() => handleSelect(item)}
          >
            {item}
          </button>
        )) : (
          <span className="pgn__form-multiselect-items-text">{noOptionsText}</span>
        )}
      </div>
    </div>
  );
});

FormMultiselect.propTypes = {
  /** Specifies the contents of the options rows. */
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  /** Specifies floating label to display for the select component. */
  floatingLabel: PropTypes.node,
  /** Specifies the default message if there is no options in the dropdown. */
  noOptionsText: PropTypes.string,
  /** Specifies selected options. */
  selectedOptions: PropTypes.arrayOf(PropTypes.string),
  /** Specifies function that changes `selectedOptions`. */
  setSelectedOptions: PropTypes.func,
  /** Informs user if there are errors. */
  errorText: PropTypes.string,
  /** Specifies whether the `Multiselect` is disabled. */
  disabled: PropTypes.bool,
  /** The `Multiselect` style variant to use. */
  variant: PropTypes.oneOf(['dark']),
};

FormMultiselect.defaultProps = {
  floatingLabel: undefined,
  noOptionsText: MULTISELECT_NO_OPTIONS_TEXT,
  selectedOptions: [],
  setSelectedOptions: () => {},
  errorText: '',
  disabled: false,
  variant: undefined,
};

export default FormMultiselect;
