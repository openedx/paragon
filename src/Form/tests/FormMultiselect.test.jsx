import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import FormMultiselect from '../FormMultiselect';

const options = [
  'White',
  'Black',
  'Green',
  'Purple',
  'Blue',
];

const Multiselect = (props) => <FormMultiselect {...props} />;

describe('<FormMultiselect />', () => {
  describe('correct rendering', () => {
    it('renders without props', () => {
      const tree = renderer.create((
        <Multiselect options={options} />
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with no options', () => {
      const wrapper = mount(<Multiselect />);
      const multiselect = wrapper.find('.pgn__form-multiselect');
      expect(multiselect.length).toEqual(1);
    });
    it('renders with the default class', () => {
      const wrapper = mount(<Multiselect options={options} />);
      const multiselect = wrapper.find('.pgn__form-multiselect');
      expect(multiselect.length).toEqual(1);
    });
    it('opens dropdown when show button is clicked', () => {
      const wrapper = mount(<Multiselect options={options} />);
      wrapper.find('.pgn__form-multiselect-field-show-btn').first().simulate('click');
      const dropdown = wrapper.find('.pgn__form-multiselect-items');
      expect(dropdown.hasClass('show')).toEqual(true);
    });
    it('options are rendered in the dropdown', () => {
      const wrapper = mount(<Multiselect options={options} />);
      const dropdownItems = wrapper.find('.pgn__form-multiselect-items');
      expect(dropdownItems.find('.pgn__form-multiselect-item')).toHaveLength(5);
    });
    it('rendered with floating label', () => {
      const labelText = 'TestMultiselect';
      const wrapper = mount(<Multiselect options={options} floatingLabel={labelText} />);
      const input = wrapper.find('.form-control');
      expect(input.at(0).props().placeholder).toEqual(labelText);
      input.simulate('focus');
      expect(wrapper.find('.pgn__form-multiselect-field-label').text()).toEqual(labelText);
    });
    it('rendered with errorText', () => {
      const errorText = 'TestError';
      const wrapper = mount(<Multiselect options={options} errorText={errorText} />);
      const error = wrapper.find('.pgn__form-multiselect-field-error');
      expect(error.text()).toEqual(errorText);
      expect(wrapper.find('.pgn__form-multiselect-field').hasClass('error')).toEqual(true);
    });
    it('rendered with disabled prop', () => {
      const wrapper = mount(<Multiselect options={options} disabled />);
      expect(wrapper.find('.pgn__form-multiselect-field').hasClass('disabled')).toEqual(true);
    });
    it('rendered with dark variant', () => {
      const wrapper = mount(<Multiselect options={options} variant="dark" />);
      expect(wrapper.find('.pgn__form-multiselect-field-label').hasClass('dark')).toEqual(true);
    });
  });
  describe('correct interactions', () => {
    it('option is selected', () => {
      const setSelectedOptions = jest.fn();
      const wrapper = mount(
        <Multiselect
          options={options}
          setSelectedOptions={setSelectedOptions}
        />,
      );
      const firstItem = wrapper.find('.pgn__form-multiselect-item').first();
      const firstItemsText = firstItem.text();
      firstItem.simulate('click');
      wrapper.update();
      expect(wrapper.find('.pgn__form-multiselect-item').first().text()).not.toEqual(firstItemsText);
      expect(setSelectedOptions).toHaveBeenCalled();
    });
    it('option is preselected', () => {
      const option = 'White';
      const wrapper = mount(
        <Multiselect
          options={options}
          selectedOptions={[option]}
        />,
      );
      expect(wrapper.find('.pgn__form-multiselect-item').first().text()).not.toEqual(option);
      expect(wrapper.find('.pgn__form-multiselect-field-chip').first().text()).toEqual(option);
    });
    it('option is removed', () => {
      const option = 'White';
      const setSelectedOptions = jest.fn();
      const wrapper = mount(
        <Multiselect
          options={options}
          selectedOptions={[option]}
          setSelectedOptions={setSelectedOptions}
        />,
      );
      const chip = wrapper.find('.pgn__form-multiselect-field-chip').first();
      chip.simulate('click');
      expect(setSelectedOptions).toHaveBeenCalled();
    });
    it('options are searched', () => {
      const searchValue = 'b';
      const wrapper = mount(
        <Multiselect
          options={options}
        />,
      );
      expect(wrapper.find('.pgn__form-multiselect-item')).toHaveLength(5);
      const input = wrapper.find('.form-control');
      input.simulate('change', { target: { value: searchValue } });
      expect(wrapper.find('.pgn__form-multiselect-item')).toHaveLength(2);
    });
    it('dropdown opens when input value is changed', () => {
      const searchValue = 'b';
      const wrapper = mount(
        <Multiselect
          options={options}
        />,
      );
      expect(wrapper.find('.pgn__form-multiselect-items').hasClass('none')).toEqual(true);
      const input = wrapper.find('.form-control');
      input.simulate('change', { target: { value: searchValue } });
      expect(wrapper.find('.pgn__form-multiselect-items').hasClass('show')).toEqual(true);
    });
    it('options are reset', () => {
      const option = 'White';
      const setSelectedOptions = jest.fn();
      const wrapper = mount(
        <Multiselect
          options={options}
          selectedOptions={[option]}
          setSelectedOptions={setSelectedOptions}
        />,
      );
      const reset = wrapper.find('.pgn__form-multiselect-field-hide-btn').first();
      reset.simulate('click');
      expect(setSelectedOptions).toHaveBeenCalled();
    });
    it('correct floating label behavior', () => {
      const labelText = 'TestMultiselect';
      const wrapper = mount(<Multiselect options={options} floatingLabel={labelText} />);
      const input = wrapper.find('.form-control');
      input.simulate('focus');
      expect(wrapper.find('.pgn__form-multiselect-field-label').text()).toEqual(labelText);
      input.simulate('blur');
      expect(input.at(0).props().placeholder).toEqual(labelText);
      wrapper.setProps({ selectedOptions: ['White'] });
      expect(wrapper.find('.pgn__form-multiselect-field-label').text()).toEqual(labelText);
    });
    it('no options text is rendered', () => {
      const searchValue = 'something';
      const noOptions = 'no options';
      const wrapper = mount(
        <Multiselect
          options={options}
          noOptionsText={noOptions}
        />,
      );
      const input = wrapper.find('.form-control');
      input.simulate('change', { target: { value: searchValue } });
      expect(wrapper.find('.pgn__form-multiselect-items-text').text()).toEqual(noOptions);
    });
  });
});
