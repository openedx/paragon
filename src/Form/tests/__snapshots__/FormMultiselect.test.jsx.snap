// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormMultiselect /> correct rendering renders without props 1`] = `
<div
  className="pgn__form-multiselect"
>
  <div
    className="pgn__form-multiselect-field"
  >
    <div
      className="pgn__form-multiselect-field-wrapper"
    >
      <div
        className="pgn__form-multiselect-field-label slide-down"
      />
      <div
        className="pgn__form-control-decorator-group pgn__form-multiselect-search"
      >
        <input
          className="form-control"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="pgn__form-multiselect-button-group"
    >
      <span
        className="pgn__form-multiselect-button-group-border"
      />
      <button
        className="pgn__form-multiselect-field-show-btn btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="pgn__icon btn-icon-after"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6-1.41-1.41z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <span
    className="pgn__form-multiselect-field-error"
  >
    
  </span>
  <div
    className="pgn__form-multiselect-items none"
  >
    <button
      className="pgn__form-multiselect-item "
      onClick={[Function]}
      type="button"
    >
      White
    </button>
    <button
      className="pgn__form-multiselect-item "
      onClick={[Function]}
      type="button"
    >
      Black
    </button>
    <button
      className="pgn__form-multiselect-item "
      onClick={[Function]}
      type="button"
    >
      Green
    </button>
    <button
      className="pgn__form-multiselect-item "
      onClick={[Function]}
      type="button"
    >
      Purple
    </button>
    <button
      className="pgn__form-multiselect-item "
      onClick={[Function]}
      type="button"
    >
      Blue
    </button>
  </div>
</div>
`;
